### ISSUE TITLE
Building error on debian

### ISSUE THREAD

**Issue by CountOmega** [sagebind/castaway#8]
I am in the process of packaging castway for debian. The build fails with an E0412:

```
  Compiling castaway v0.2.1 (/usr/share/cargo/registry/castaway-0.2.1)
     Running `CARGO=/usr/bin/cargo CARGO_CRATE_NAME=castaway CARGO_MANIFEST_DIR=/usr/share/cargo/registry/castaway-0.2.1 CARGO_PKG_AUTHORS='Stephen M. Coakley <me@stephencoakley.com>' CARGO_PKG_DESCRIPTION='Safe, zero-cost downcasting for limited compile-time specialization.' CARGO_PKG_HOMEPAGE='' CARGO_PKG_LICENSE=MIT CARGO_PKG_LICENSE_FILE='' CARGO_PKG_NAME=castaway CARGO_PKG_REPOSITORY='https://github.com/sagebind/castaway' CARGO_PKG_VERSION=0.2.1 CARGO_PKG_VERSION_MAJOR=0 CARGO_PKG_VERSION_MINOR=2 CARGO_PKG_VERSION_PATCH=1 CARGO_PKG_VERSION_PRE='' CARGO_PRIMARY_PACKAGE=1 LD_LIBRARY_PATH='/tmp/tmp.wJuFF5GI1P/target/debug/deps:/usr/lib' rustc --crate-name castaway --edition=2018 src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 --test -C metadata=f06c949c70501ccf -C extra-filename=-f06c949c70501ccf --out-dir /tmp/tmp.wJuFF5GI1P/target/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -C incremental=/tmp/tmp.wJuFF5GI1P/target/x86_64-unknown-linux-gnu/debug/incremental -L dependency=/tmp/tmp.wJuFF5GI1P/target/x86_64-unknown-linux-gnu/debug/deps -L dependency=/tmp/tmp.wJuFF5GI1P/target/debug/deps --extern paste=/tmp/tmp.wJuFF5GI1P/target/debug/deps/libpaste-9b43ddbe3bcf5cdd.so --extern rustversion=/tmp/tmp.wJuFF5GI1P/target/debug/deps/librustversion-44ce5c41119c978a.so -C debuginfo=2 --cap-lints warn -C linker=x86_64-linux-gnu-gcc -C link-arg=-Wl,-z,relro --remap-path-prefix /usr/share/cargo/registry/castaway-0.2.1=/usr/share/cargo/registry/castaway-0.2.1`
error[E0433]: failed to resolve: use of undeclared type `String`
   --> src/lib.rs:482:13
    |
482 |             String::from("hello world") => Ok(ref v) if v.as_str() == "hello world",
    |             ^^^^^^ use of undeclared type `String`

error[E0412]: cannot find type `String` in this scope
   --> src/lib.rs:481:13
    |
481 |         for String {
    |             ^^^^^^ not found in this scope

Some errors have detailed explanations: E0412, E0433.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `castaway` due to 2 previous errors
```
Compiler: GCC 11

**Comment by NobodyXu** [comment]
What is the command you used to build `castaway`?

**Comment by CountOmega** [comment]
sbuild, I'm building in a chroot. It could be just a broken test, but I think this is an error.

**Comment by NobodyXu** [comment]
> cargo 0.57.0

Do you mean cargo 1.57.0?

**Comment by CountOmega** [comment]
Sorry, I meant 1.56.0

**Comment by NobodyXu** [comment]
> use of undeclared type `String`

Did you enable `no_std` support?

If you do, then the test in this library wouldn't compile.

What `cargo` command are you using to build this library?

**Comment by CountOmega** [comment]
That one:
```
autopkgtest [13:04:11]: test librust-castaway-dev:: /usr/share/cargo/bin/cargo-auto-test castaway 0.2.1 --all-targets --no-default-features
autopkgtest [13:04:11]: test librust-castaway-dev:: [-----------------------
debian cargo wrapper: options, profiles, parallel: ['parallel=8'] [] ['-j8']
debian cargo wrapper: rust_type, gnu_type: x86_64-unknown-linux-gnu, x86_64-linux-gnu
debian cargo wrapper: linking /usr/share/cargo/registry/* into /tmp/tmp.wJuFF5GI1P/registry/
debian cargo wrapper: options, profiles, parallel: ['parallel=8'] [] ['-j8']
debian cargo wrapper: rust_type, gnu_type: x86_64-unknown-linux-gnu, x86_64-linux-gnu
debian cargo wrapper: running subprocess (['env', 'RUST_BACKTRACE=1', '/usr/bin/cargo', '-Zavoid-dev-deps', 'test', '--verbose', '--verbose', '-j8', '--target', 'x86_64-unknown-linux-gnu', '--all-targets', '--no-default-features'],) {}
```

**Comment by NobodyXu** [comment]
You need to either remove `--all-targets` so that the testing code won't be compiled, or remove `--no-default-features` so that std library is enabled.

**Comment by sagebind** [comment]
This is happening because the `std` crate feature is not enabled (which it is by default, but is disabled via `--no-default-features`), but the test attempting to be compiled expects it to be enabled. I have a PR here to make these tests conditional to help: https://github.com/sagebind/castaway/pull/9

**Comment by sagebind** [comment]
OK, I've released version 0.2.2 which should correct this problem. Though if you must still package version 0.2.1 then the solution is to either (a) don't compile the tests, or (b) don't disable support for the standard library.
